Plugins: additional settings should know about the node settings.

martijnvg opened the pull request:
Pass down the node settings to the Plugin#addtionalSettings(...) method.

martijnvg committed:
Plugins: Pass down the node settings to the Plugin#addtionalSettings(...) method.

dadoonet commented on src/main/java/org/elasticsearch/plugins/Plugin.java:
```diff
@@ -100,6 +100,8 @@
+     * @param settings The already configured node settings
      */
-    Settings additionalSettings();
+    Settings additionalSettings(Settings settings);
```
This will break all plugins and will force all authors to update their plugins.
Not sure if there are any other option but if not we should mark it as a breaking change.

Can we add it only to `AbstractPlugin`? I mean add it as a new function but not replacing the existing signature.

dadoonet commented on src/main/java/org/elasticsearch/plugins/AbstractPlugin.java:
```diff
@@ -115,7 +115,7 @@ public void processModule(Module module) {
     @Override
-    public Settings additionalSettings() {
+    public Settings additionalSettings(Settings settings) {
         return ImmutableSettings.Builder.EMPTY_SETTINGS;
```
`settings` is not used in that case?

martijnvg replied:
Nope, because it is about the additional settings a plugin can set. This just means that by default no additional settings are added to a node's settings.

dadoonet commented:
@martijnvg left some comments

martijnvg commented:
@dadoonet Good point about the breaking part, maybe I was too eager with that. I think for 1.x it makes sense to not make the change in signature in the Plugin interface. Let me change this.

I think for 2.0 this change as it is know is okay?

dadoonet commented:
For 2.0, I'm +1 for breaking the compatibility for `Plugin` interface if we need to.

martijnvg committed:
Make sure that the change isn't breaking:
Removed the breaking change in the Plugin interface and added
Added the additional method to AbstractPlugin in non breaking way that by default delegates the #additionalSettings()

martijnvg commented:
@dadoonet I updated the PR to be non breaking.

uboness commented:
@martijnvg maybe I'm missing something... why do we need this support when node settings can be injected to the plugin in the ctor?

martijnvg commented:
@uboness I did not know that a plugin can either have a constructor with Settings arg or a no argument constructor! I thought it was only a no argument constructor. ( PluginService#loadPlugin(...) )

Bringing the node settings to a Plugin implementation is what the purposes of this PR... so it can be closed.
